lib/onelogin/ruby-saml/response.rb

```diff
@@ -112,78 +151,326 @@ def status_message
+      # @return [Boolean] True if the required info is found, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_response_state(soft = true)
+        if response.nil? || response.empty?
+          @errors << "Blank SAML Response"
+          return soft ? false : validation_error("Blank SAML Response")
+        end
+
+        if settings.nil?
+          @errors << "No settings on SAML Response"
+          return soft ? false : validation_error("No settings on SAML Response")
+        end
+
+        if settings.idp_cert_fingerprint.nil? && settings.idp_cert.nil?
+          @errors << "No fingerprint or certificate on settings"
+          return soft ? false : validation_error("No fingerprint or certificate on settings")
+        end
+
+        true
+      end
+
+      # Validates that the SAML Response contains an ID 
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response contains an ID, otherwise returns False
+      #
+      def validate_id()
+        unless id(document)
+          @errors << "Missing ID attribute on SAML Response"
+          return false
+        end
+        true
+      end
+
+      # Validates the SAML version (2.0)
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response is 2.0, otherwise returns False
+      #
+      def validate_version()
+        unless version(document) == "2.0"
+          @errors << "Unsupported SAML version"
+          return false
+        end
+        true
+      end
+
+      # Validates that the SAML Response only contains a single Assertion (encrypted or not).
+      # If fails, the error is added to the errors array.
+      # @return [Boolean] True if the SAML Response contains one unique Assertion, otherwise False
+      #
+      def validate_num_assertion()
+        assertions = REXML::XPath.match(document, "//a:Assertion", { "a" => ASSERTION })
+        encrypted_assertions = REXML::XPath.match(document, "//a:EncryptedAssertion", { "a" => ASSERTION })
+
+        unless assertions.size + encrypted_assertions.size == 1
+          @errors << "SAML Response must contain 1 assertion"
+          return false
+        end
+        true
+      end
+
+      # Validates the Status of the SAML Response
+      # If fails, the error is added to the errors array, including the StatusCode returned and the Status Message.
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)      
+      # @return [Boolean] True if the SAML Response contains a Success code, otherwise False if soft == false
+      # @raise [ValidationError] if soft == false and validation fails
+      #
       def validate_success_status(soft = true)
```

Comment:
This method is completely duplicated in multiple places. It would make a lot more sense to move these to a validation module and share the logic.